File: examples/physics/physx_h1_ragdoll.cpp
///
/// H1 Robot — Ragdoll (floating base) + Active PD Control
///
/// Identical to physx_h1.cpp except the articulation root is NOT fixed:
///   artic->setArticulationFlag(PxArticulationFlag::eFIX_BASE, false)
///
/// The robot spawns above ground and falls freely.
/// With kp=0, kd=small → pure ragdoll (limp, gravity only).
/// With kp>0               → active ragdoll (tries to hold pose while falling).
///
/// Controls:
///   Space : pause / resume
///   R     : reset robot at initial height
///   WASD + mouse : camera
///   ImGui : PD gains, per-joint target angles, spawn height
///

#include "kangEngine.hpp"
#include "physics/force_drag_controller.hpp"
#include <glm/glm.hpp>
#include <glm/geometric.hpp>
#include <imgui.h>
#include <algorithm>
#include <memory>
#include <string>
#include <vector>

using namespace KE;
using namespace KE::Asset;
using namespace KE::Animation;
using namespace KE::Bridge;
using namespace physx;

// ---------------------------------------------------------------------------
// App
// ---------------------------------------------------------------------------

class H1RagdollApp : public App {
  public:
    std::unique_ptr<Backend::Shader> commonShader;
    std::unique_ptr<Backend::Shader> groundShader;

    SkeletonBridge robot;
    PhysicsWorld physics{PhysicsConfig::zUp()};
    Articulation artic;
    Bridge::PhysicsBridge physicsBridge;
    ForceDragController forceDrag;
    ForceDragConfig forceDragConfig;

    std::vector<float> targets;
    std::vector<RenderableHandle> bodyHandles;
    RenderableHandle contactArrowHandle = InvalidHandle;

    float spawnHeightOffset = 1.5f;
    float contactImpulseScale = 0.08f;
    float contactMaxArrowLength = 0.6f;
    float dragForceArrowScale = 0.003f;

    float kp = 0.f;
    float kd = 5.f;
    bool paused = false;
    bool spaceWasDown = false;
    bool rWasDown = false;
    bool showCollision = false;
    bool showContacts = true;
    bool showDragForceArrow = true;
    // -----------------------------------------------------------------------
    void setup() override {
        auto commonVSPath = KE::getAssetPath("shaders/common.vs");
        auto commonFSPath = KE::getAssetPath("shaders/common.fs");
        auto groundFSPath = KE::getAssetPath("shaders/checkerboard.fs");

        commonShader = getRenderer().device()->createShaderFromFile(
            commonVSPath, commonFSPath);
        groundShader = getRenderer().device()->createShaderFromFile(
            commonVSPath, groundFSPath);

        groundShader->use();
        groundShader->setUniformBlockBinding("cameraUBO", 0);
        groundShader->setUniformBlockBinding("lightUBO", 1);
        auto white = ColorLibrary::get(KE::ColorType::WHITE);
        auto pG = ColorLibrary::get(KE::ColorType::PASTEL_GREEN);
        groundShader->setVec4("checkerColor1",
                              glm::vec4(white.r, white.g, white.b, white.a));
        groundShader->setVec4("checkerColor2",
                              glm::vec4(pG.r, pG.g, pG.b, pG.a));
        commonShader->use();
        commonShader->setUniformBlockBinding("cameraUBO", 0);
        commonShader->setUniformBlockBinding("lightUBO", 1);

        std::string basePath = "external/skybox";
        setSkybox(KE::getAssetPath(basePath + "/Cubemap_Sky_08-512x512.png")
                  /*{
                    KE::getAssetPath(basePath + "/right.jpg"),
                    KE::getAssetPath(basePath + "/left.jpg"),
                    KE::getAssetPath(basePath + "/top.jpg"),
                    KE::getAssetPath(basePath + "/bottom.jpg"),
                    KE::getAssetPath(basePath + "/front.jpg"),
                    KE::getAssetPath(basePath + "/back.jpg"),
                  },*/);

        physics.addDefaultGround();
        auto* gnd = getScene()->definePrim("/ground", Scene::PrimType::Mesh);
        gnd->setMeshData(std::make_shared<Scene::MeshData>(
            Scene::Prim::createPlaneData(100.f, UpAxis::Z)));
        addRenderable(groundShader.get(), gnd);

        // const std::string mjcfPath =
        //     KE::getAssetPath("external/retargetted/unitree_h1/unitree_h1.xml");
        const std::string mjcfPath = KE::getAssetPath("characters/kw/kw5.xml");
        const auto mjcfData = MJCFLoader::load(mjcfPath);

        artic = Articulation::build(physics, mjcfData.skeletonTree,
                                    mjcfData.collisionGeoms, mjcfData.joints,
                                    mjcfData.inertials,
                                    ArticulationConfig::freeBase());

        robot = SkeletonBridge::fromData(mjcfData, getScene());
        physicsBridge.add(artic, robot);
        bodyHandles.clear();
        bodyHandles.reserve(robot.bodyPrims().size());
        for (auto* prim : robot.bodyPrims())
            bodyHandles.push_back(addRenderable(commonShader.get(), prim));
        forceDrag.registerArticulation(artic, bodyHandles);

        auto colPrims = physicsBridge.addCollisionVisuals(artic, getScene());
        for (auto* p : colPrims)
            addRenderable(commonShader.get(), p);

        contactArrowHandle = Scene::DebugDraw::logArrows(
            this, commonShader.get(), "/debug/contact_arrows",
            {glm::vec3(0.0f)}, {glm::vec3(0.0f, 0.0f, 0.1f)},
            {glm::vec4(1.0f, 0.2f, 0.05f, 1.0f)}, 0.025f, 12);

        targets.assign(artic.numDofs(), 0.f);

        reset();
        checkError();
    }

    void onForceDragBegin(const RayPickResult& result,
                          const glm::vec3& target) override {
        forceDrag.begin(result, target);
        updateDragForceArrow();
    }

    void onForceDragUpdate(const RayPickResult&,
                           const glm::vec3& target) override {
        forceDrag.update(target);
        updateDragForceArrow();
    }

    void onForceDragEnd() override {
        forceDrag.end();
        clearDebugLines("/debug/drag_force");
        clearDebugPoints("/debug/drag_force_target");
    }

    void updateDragForceArrow() {
        if (!showDragForceArrow || !forceDrag.active()) {
            clearDebugLines("/debug/drag_force");
            clearDebugPoints("/debug/drag_force_target");
            return;
        }

        const glm::vec3 start = forceDrag.lastAnchorPosition();
        const glm::vec3 force = forceDrag.lastForce();
        const float forceLen = glm::length(force);
        if (forceLen < 1e-5f) {
            clearDebugLines("/debug/drag_force");
            return;
        }

        const glm::vec3 dir = force / forceLen;
        const glm::vec3 end = start + force * dragForceArrowScale;
        const float shaftLen = glm::length(end - start);
        const float headLen = std::clamp(shaftLen * 0.25f, 0.04f, 0.18f);

        glm::vec3 side = glm::cross(dir, glm::vec3(0.0f, 0.0f, 1.0f));
        if (glm::length(side) < 1e-5f)
            side = glm::cross(dir, glm::vec3(0.0f, 1.0f, 0.0f));
        side = glm::normalize(side);

        const glm::vec3 back = end - dir * headLen;
        const glm::vec3 left = back + side * headLen * 0.45f;
        const glm::vec3 right = back - side * headLen * 0.45f;
        const auto red = ColorLibrary::get(ColorType::RED);
        const glm::vec4 arrowColor(red.r, red.g, red.b, 1.0f);
        logDebugLines("/debug/drag_force", {start, end, end},
                      {end, left, right}, {arrowColor}, 3.0f);

        const auto mag = ColorLibrary::get(ColorType::MAGENTA);
        logDebugPoints("/debug/drag_force_target", {forceDrag.lastTarget()},
                       {glm::vec4(mag.r, mag.g, mag.b, 1.0f)}, 10.0f);
    }

    void reset() {
        artic.resetRoot(PxTransform(PxVec3(0.f, 0.f, spawnHeightOffset),
                                    PxQuat(0.0f, 0.0f, 0.0f, 1.0f)));
    }

    // -----------------------------------------------------------------------
    void preRender() override {
        bool spaceDown = glfwGetKey(getWindow(), GLFW_KEY_SPACE) == GLFW_PRESS;
        if (spaceDown && !spaceWasDown)
            paused = !paused;
        spaceWasDown = spaceDown;

        bool rDown = glfwGetKey(getWindow(), GLFW_KEY_R) == GLFW_PRESS;
        if (rDown && !rWasDown)
            reset();
        rWasDown = rDown;

        if (!paused) {
            artic.setDriveTargets(targets, kp, kd);
            physics.step();
            physicsBridge.sync();
        }
        updateContactArrows();
        updateDragForceArrow();

        checkError();
    }

    void updateContactArrows() {
        if (contactArrowHandle == InvalidHandle)
            return;

        std::vector<glm::mat4> transforms;
        std::vector<glm::vec4> colors;
        if (showContacts) {
            const auto& contacts = physics.getContacts();
            transforms.reserve(contacts.size());
            colors.reserve(contacts.size());

            for (const auto& contact : contacts) {
                float impulseLen = glm::length(contact.impulse);
                glm::vec3 dir = impulseLen > 1e-5f
                                    ? glm::normalize(contact.impulse)
                                    : contact.normal;
                if (glm::length(dir) < 1e-5f)
                    continue;

                float arrowLen = std::clamp(impulseLen * contactImpulseScale,
                                            0.05f, contactMaxArrowLength);
                glm::vec3 start = contact.position;
                glm::vec3 end = start + dir * arrowLen;
                glm::mat4 transform(1.0f);
                if (!Scene::DebugDraw::makeArrowTransform(start, end,
                                                          transform))
                    continue;
                transforms.push_back(transform);

                float heat =
                    std::clamp(arrowLen / contactMaxArrowLength, 0.0f, 1.0f);
                colors.push_back(
                    glm::vec4(1.0f, 0.9f - 0.7f * heat, 0.05f, 1.0f));
            }
        }
        updateRenderableTransforms(contactArrowHandle, transforms, &colors);
    }

    // -----------------------------------------------------------------------
    void render() override {
        int n = artic.numLinks();

        ImGui::Begin("H1 Ragdoll");
        ImGui::Text("Bodies: %d  DOFs: %d  |  %s", n, artic.numDofs(),
                    paused ? "PAUSED" : "running");
        ImGui::Text("Space: pause/resume    R: reset");
        ImGui::Separator();

        ImGui::Text("Mode: %s",
                    kp < 1.f ? "Pure ragdoll (kp=0)" : "Active ragdoll (kp>0)");
        ImGui::SliderFloat("kp (stiffness — 0=limp)", &kp, 0.f, 500.f);
        ImGui::SliderFloat("kd (damping)", &kd, 0.f, 50.f);
        ImGui::SliderFloat("Spawn height offset (m)", &spawnHeightOffset, 0.5f,
                           5.f);
        ImGui::Separator();

        if (ImGui::Button("Reset targets to zero"))
            std::fill(targets.begin(), targets.end(), 0.f);
        ImGui::SameLine();
        if (ImGui::Button("Reset"))
            reset();
        ImGui::SameLine();
        if (ImGui::Checkbox("Show collision", &showCollision)) {
            physicsBridge.setCollisionVisible(showCollision);
            float alpha = showCollision ? 0.12f : 1.0f;
            for (auto* p : robot.bodyPrims()) {
                auto col = p->getDisplayColorAlpha().value_or(
                    glm::vec4(0.15f, 0.15f, 0.15f, 1.f));
                col.a = alpha;
                p->setDisplayColorAlpha(col);
            }
        }
        ImGui::Checkbox("Show contacts", &showContacts);
        ImGui::SliderFloat("Contact impulse scale", &contactImpulseScale,
                           0.005f, 0.3f);
        ImGui::SliderFloat("Contact max arrow length", &contactMaxArrowLength,
                           0.05f, 2.0f);
        ImGui::Text("Contacts: %u", physics.numContacts());
        ImGui::Separator();
        ImGui::Text("Force drag: select Force mode, then Shift + left drag");
        ImGui::Checkbox("Show drag force arrow", &showDragForceArrow);
        if (ImGui::SliderFloat("Drag stiffness", &forceDragConfig.stiffness,
                               0.0f, 1000.0f) ||
            ImGui::SliderFloat("Drag damping", &forceDragConfig.damping, 0.0f,
                               80.0f) ||
            ImGui::SliderFloat("Drag max force", &forceDragConfig.maxForce,
                               0.0f, 1500.0f)) {
            forceDrag.setConfig(forceDragConfig);
        }
        ImGui::SliderFloat("Drag arrow scale", &dragForceArrowScale, 0.0005f,
                           0.02f);
        ImGui::Text("Dragging: %s", forceDrag.active() ? "yes" : "no");

        if (kp >= 1.f) {
            ImGui::Text("Joint targets (rad):");
            ImGui::BeginChild("joints", ImVec2(0, 350), true);
            int dofIdx = 0;
            for (int i = 1; i < n; i++) {
                auto jit = artic.joints().find(i);
                if (jit == artic.joints().end())
                    continue;
                for (const auto& jd : jit->second)
                    ImGui::SliderFloat(jd.name.c_str(), &targets[dofIdx++],
                                       jd.loLimit, jd.hiLimit);
            }
            ImGui::EndChild();
        } else {
            ImGui::TextDisabled("(Set kp > 0 to enable joint targets)");
        }
        ImGui::End();

        ImGui::Begin("Visualize Torque");
        ImGui::Text("Real-time Joint Torques (Nm):");
        ImGui::Separator();

        PxArticulationCache* cache = artic.raw()->createCache();
        artic.raw()->copyInternalStateToCache(
            *cache, PxArticulationCacheFlag::eJOINT_SOLVER_FORCES |
                        PxArticulationCacheFlag::eFORCE |
                        PxArticulationCacheFlag::ePOSITION |
                        PxArticulationCacheFlag::eVELOCITY);

        ImGui::BeginChild("torques", ImVec2(0, 300), true);
        {
            int dofIdx = 0;
            for (int i = 1; i < n; i++) {
                auto jit = artic.joints().find(i);
                if (jit == artic.joints().end())
                    continue;
                for (const auto& jd : jit->second) {
                    // Is this same with physx's implicit pd control?
                    float currentPos = cache->jointPosition[dofIdx];
                    float currentVel = cache->jointVelocity[dofIdx];
                    float torque = kp * (targets[dofIdx] - currentPos) +
                                   kd * (0.f - currentVel);
                    float bar = (torque + 100.f) / 200.f;
                    bar = std::clamp(bar, 0.f, 1.f);
                    char overlay[64];
                    snprintf(overlay, sizeof(overlay), "%+.1f Nm", torque);
                    ImGui::Text("%s", jd.name.c_str());
                    ImGui::SameLine();
                    ImGui::ProgressBar(bar, ImVec2(-1, 0), overlay);
                    dofIdx++;
                }
            }
        }
        cache->release();
        ImGui::EndChild();
        ImGui::End();
    }
};

// ---------------------------------------------------------------------------
int main() {
    H1RagdollApp app;
    app.initialize(1920, 1080, false, UpAxis::Z);
    app.start();
    return 0;
}

File: python/examples/physx_h1_ragdoll.py
"""
H1 Robot Ragdoll — Python equivalent of examples/physics/physx_h1_ragdoll.cpp.

Loads an MJCF robot, builds a free-base PhysX articulation, and lets it fall
under gravity.  With kp=0 and small kd this behaves as a limp ragdoll; increase
kp in code for active pose holding.
"""

import os

import kangengine as ke
from kangengine import animation, imgui, keys, scene


def asset_path(*parts):
    base = os.path.join(os.path.dirname(ke.__file__), "assets")
    return os.path.join(base, *parts)


class H1RagdollApp(ke.App):
    def setup(self):
        self.spawn_height_offset = 1.5
        self.kp = 0.0
        self.kd = 5.0
        self.paused = False
        self.show_collision = False

        device = self.getGraphicsDevice()

        vs = asset_path("shaders", "common.vs")
        fs = asset_path("shaders", "common.fs")
        checker_fs = asset_path("shaders", "checkerboard.fs")

        self.robot_shader = device.createShaderFromFile(vs, fs)
        self.ground_shader = device.createShaderFromFile(vs, checker_fs)

        for shader in (self.robot_shader, self.ground_shader):
            shader.use()
            shader.setUniformBlockBinding("cameraUBO", 0)
            shader.setUniformBlockBinding("lightUBO", 1)

        self.ground_shader.use()
        self.ground_shader.setVec4("checkerColor1", ke.vec4(1.0, 1.0, 1.0, 1.0))
        self.ground_shader.setVec4("checkerColor2", ke.vec4(0.77, 0.93, 0.78, 1.0))

        self.physics = ke.PhysicsWorld(ke.PhysicsConfig.z_up())
        self.physics.add_default_ground()

        gnd = self.getScene().define_prim("/ground", scene.PrimType.Mesh)
        gnd.set_mesh_data(scene.Prim.create_plane_data(100.0, ke.UpAxis.Z))
        self.addShape(self.ground_shader, gnd)

        mjcf = asset_path("external", "retargetted", "kw", "kw5.xml")
        mjcf_data = animation.MJCFLoader.load(mjcf)

        self.articulation = ke.Articulation.build(
            self.physics,
            mjcf_data,
            ke.ArticulationConfig.free_base(),
        )

        self.robot = animation.SkeletonBridge.from_mjcf(
            mjcf,
            self.getScene(),
            "/robot",
            1.0,
            "DFS",
        )

        self.physics_bridge = ke.PhysicsBridge(self)
        self.physics_bridge.add(self.articulation, self.robot)

        for prim in self.robot.body_prims():
            self.addShape(self.robot_shader, prim)

        collision_prims = self.physics_bridge.add_collision_visuals(
            self.articulation,
            self.getScene(),
            "/collision",
            self.show_collision,
        )
        for prim in collision_prims:
            self.addShape(self.robot_shader, prim)

        self.targets = [0.0] * self.articulation.num_dofs()
        self.reset()

        print(
            "Ragdoll loaded: "
            f"{self.articulation.num_links()} links, "
            f"{self.articulation.num_dofs()} DOFs"
        )
        print("Defaults: kp=0.0, kd=5.0, free base, Z-up")
        self.checkError()

    def reset(self):
        self.articulation.reset_root(
            ke.vec3(0.0, 0.0, self.spawn_height_offset),
            ke.quat(1.0, 0.0, 0.0, 0.0),
        )

    def preRender(self):
        if self.was_key_pressed(keys.SPACE):
            self.paused = not self.paused

        if self.was_key_pressed(keys.R):
            self.reset()

        if self.paused:
            return

        self.articulation.set_drive_targets(self.targets, self.kp, self.kd)
        self.physics.step()
        self.physics_bridge.sync()
        self.checkError()

    def render(self):
        num_links = self.articulation.num_links()
        num_dofs = self.articulation.num_dofs()
        state = "PAUSED" if self.paused else "running"

        imgui.begin("H1 Ragdoll")
        imgui.text(f"Bodies: {num_links}  DOFs: {num_dofs}  |  {state}")
        imgui.text("Space: pause/resume    R: reset")
        imgui.separator()

        mode = "Pure ragdoll (kp=0)" if self.kp < 1.0 else "Active ragdoll (kp>0)"
        imgui.text(f"Mode: {mode}")
        _, self.kp = imgui.slider_float("kp (stiffness, 0=limp)", self.kp, 0.0, 500.0)
        _, self.kd = imgui.slider_float("kd (damping)", self.kd, 0.0, 50.0)
        _, self.spawn_height_offset = imgui.slider_float(
            "Spawn height offset (m)",
            self.spawn_height_offset,
            0.5,
            5.0,
        )
        imgui.separator()

        if imgui.button("Reset targets to zero"):
            self.targets = [0.0] * len(self.targets)
        imgui.same_line()
        if imgui.button("Reset"):
            self.reset()
        imgui.same_line()
        changed, self.show_collision = imgui.checkbox(
            "Show collision",
            self.show_collision,
        )
        if changed:
            self.physics_bridge.set_collision_visible(self.show_collision)
            self._set_robot_alpha(0.12 if self.show_collision else 1.0)

        if self.kp >= 1.0:
            imgui.text("Joint targets (rad):")
            imgui.begin_child("joints", 0.0, 350.0, True)
            dof_idx = 0
            joints = self.articulation.joints()
            for link_idx in range(1, num_links):
                for joint in joints.get(link_idx, []):
                    _, self.targets[dof_idx] = imgui.slider_float(
                        joint.name,
                        self.targets[dof_idx],
                        joint.lo_limit,
                        joint.hi_limit,
                    )
                    dof_idx += 1
            imgui.end_child()
        else:
            imgui.text_disabled("(Set kp > 0 to enable joint targets)")

        imgui.end()

    def _set_robot_alpha(self, alpha):
        for prim in self.robot.body_prims():
            color = prim.get_display_color_alpha()
            if color is None:
                color = ke.vec4(0.15, 0.15, 0.15, 1.0)
            prim.set_display_color_alpha(ke.vec4(color.x, color.y, color.z, alpha))

    def postRender(self):
        pass

    def cleanup(self):
        if hasattr(self, "physics_bridge"):
            self.physics_bridge = None
        if hasattr(self, "articulation"):
            self.articulation.release()
            self.articulation = None
        if hasattr(self, "physics"):
            self.physics = None


if __name__ == "__main__":
    missing = [
        name
        for name in (
            "PhysicsConfig",
            "PhysicsWorld",
            "ArticulationConfig",
            "Articulation",
            "PhysicsBridge",
        )
        if not hasattr(ke, name)
    ]
    if missing:
        raise RuntimeError(
            "This example requires KangEngine to be built with PhysX bindings. "
            f"Missing: {', '.join(missing)}"
        )

    app = H1RagdollApp()
    app.initialize(1920, 1080, False, ke.UpAxis.Z)
    app.start()
